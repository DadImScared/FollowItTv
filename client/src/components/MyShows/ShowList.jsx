
import React, { Component } from 'react';
import { connect } from 'react-redux';

import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Button from '@material-ui/core/Button';

import { findFollowedShows } from '../../actions/followedShows';
import { makeFollowedShowData } from '../../reducers/followedShows';

import Loading from '../Loading';

class ShowList extends Component {

  componentDidMount() {
    const {  isLoading, getFollowedShows } = this.props;
    // only make api request if loading state has not been initialized
    // this is because any additional data after the load will come from the client when following/un following a show
    if (isLoading !== undefined) return;
    getFollowedShows();
  }

  render() {
    const { unFollow, isCached, isLoading, followedShows } = this.props;

    if (isLoading) return <Loading/>;
    if (!isCached) return <div>no results</div>;

    return (
      <List>
        {
          followedShows.map((show, index) => (
            <ListItem divider={true} key={`${show.id}-${index}`}>
              <ListItemText primary={show.name} />
              <Button onClick={() => unFollow(show)}>
                delete
              </Button>
            </ListItem>
          ))
        }
      </List>
    );
  }
}

const makeMapStateToProps = () => {
  const getShows = makeFollowedShowData();
  return (state, props) => {
    return getShows(state, props);
  };
};

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...ownProps,
  ...stateProps,
  getFollowedShows: () => dispatchProps.dispatch(findFollowedShows(ownProps.day))
});

export default connect(makeMapStateToProps, null, mergeProps)(ShowList);
