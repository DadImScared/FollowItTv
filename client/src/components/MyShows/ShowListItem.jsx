
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import Button from '@material-ui/core/Button';
import ListItem from '@material-ui/core/ListItem';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import ListItemText from '@material-ui/core/ListItemText';

class ShowListItem extends React.PureComponent {

  unFollow = () => {
    const { show } = this.props;
    this.props.unFollow(show);
  };

  render() {
    const { show } = this.props;
    return (
      <ListItem component={Link} to={`/show/${show.id}`} button divider={true}>
        <ListItemText style={{ flex: '0.7 1 auto' }} primary={show.name} />
        <ListItemSecondaryAction>
          <Button onClick={this.unFollow}>
            delete
          </Button>
        </ListItemSecondaryAction>
      </ListItem>
    );
  }
}

ShowListItem.propTypes = {
  show: PropTypes.object.isRequired,
  unFollow: PropTypes.func.isRequired
};

export default ShowListItem;
