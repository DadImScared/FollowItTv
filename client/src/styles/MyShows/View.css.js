
export default () => ({
  flexContainer: {
    [`@media only screen and (min-width: 1744px)`]: {
      justifyContent: 'center'
    }
  }
});
