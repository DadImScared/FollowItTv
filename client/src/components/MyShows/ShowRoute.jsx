
import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import PropTypes from 'prop-types';

import ShowList from './ShowList';

class ShowRoute extends Component {
  route = () => {
    const { day, unFollow } = this.props;
    return (<ShowList day={day} unFollow={unFollow} />);
  };

  render() {
    const { day = 'All', match } = this.props;
    return (
      <Route
        exact
        path={
          day === 'All' ?
            `(${match.url}|${match.url}/All)`
            :
            `${match.url}/${day}`
        }
        render={this.route}
      />
    );
  }
}

ShowRoute.propTypes = {
  day: PropTypes.string.isRequired,
  unFollow: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired
};

export default ShowRoute;
