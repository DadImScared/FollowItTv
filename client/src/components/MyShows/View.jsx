
import React from 'react';

import { Route } from 'react-router-dom';

import SwipeableViews from 'react-swipeable-views';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Snackbar from '@material-ui/core/Snackbar';
import Button from '@material-ui/core/Button';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

import HideNav from '../HideNav';
import ShowList from './ShowList';
import { View as styles } from '../../styles/MyShows';

const View = ({
  classes,
  days,
  handleChange,
  handleSwipeChange,
  day,
  match,
  isOpen,
  undoData: { show = {}, key },
  unFollow,
  undoAction,
  handleClose,
  directionDown,
  handleSnackbarExit
}) => (
  <Paper style={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
    <HideNav directionDown={directionDown}>
      <Tabs
        scrollable
        scrollButtons="auto"
        value={day}
        onChange={handleChange}
        classes={{
          flexContainer: classes.flexContainer
        }}
      >
        {
          days.map((item, index) => (
            <Tab label={item} key={`${item}-${index}`} />
          ))
        }
      </Tabs>
    </HideNav>
    <SwipeableViews
      style={{ flexGrow: 1 }}
      index={day}
      onChangeIndex={handleSwipeChange}
    >
      {
        days.map((item, index) => (
          <Route
            exact
            path={
              item === 'All' ?
                `(${match.url}|${match.url}/All)`
                :
                `${match.url}/${item}`
            }
            render={() => <ShowList day={item} unFollow={unFollow}  />}
            key={`${item}-${index}-route`}
          />
        ))
      }
    </SwipeableViews>
    <Snackbar
      key={key}
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left'
      }}
      open={isOpen}
      onClose={handleClose}
      onExited={handleSnackbarExit}
      autoHideDuration={6000}
      message={`Un followed ${show.name}`}
      action={(
        <Button onClick={undoAction}>
          Undo
        </Button>
      )}
    />
  </Paper>
);

export default withStyles(styles)(View);
