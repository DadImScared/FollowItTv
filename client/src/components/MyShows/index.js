
import React, { Component } from 'react';

import moment from 'moment';
import { connect } from 'react-redux';

import {
  createFollowedShow,
  deleteFollowedShow
} from '../../actions/followedShows';
import View from './View';


const routes = {
  0: 'All',
  1: 'Monday',
  2: 'Tuesday',
  3: 'Wednesday',
  4: 'Thursday',
  5: 'Friday',
  6: 'Saturday',
  7: 'Sunday'
};

export class MyShows extends Component {
  queue = [];

  constructor(...args) {
    super(...args);
    this.state = {
      day: 0,
      undoData: {},
      isOpen: false
    };
  }

  componentDidMount() {
    this.setInitialTab(this.props);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { location: { pathname: oldPath } } = this.props;
    const { location: { pathname } } = nextProps;
    if (pathname !== oldPath) {
      this.setInitialTab(nextProps);
    }
  }

  handleChange = (event, value) => {
    this.pushRoute(value);
  };

  handleSwipeChange = (value) => {
    this.pushRoute(value);
  };

  pushRoute = (value) => {
    if (value === this.state.day) {
      return;
    }
    const { match, history } = this.props;
    this.setState({ day : value });
    history.push(`${match.url}/${routes[value]}`);
  };

  setInitialTab = (props) => {
    const { location } = props;
    for(const route of Object.keys(routes)) {
      if (location.pathname.includes(routes[route])) {
        this.setState({ day: parseInt(route) });
        return;
      }
    }
  };

  unFollow = async (show) => {
    const { schedule: { days }, id } = show;
    const { dispatch } = this.props;
    this.setUndoData(show);
    await dispatch(deleteFollowedShow(days, id));
    if (this.state.isOpen) {
      this.clearSnackbar();
    }
    else {
      this.processQueue();
    }
  };

  undoAction = async () => {
    const { undoData: { show: { schedule: { days }, id } } } = this.state;
    const { dispatch } = this.props;
    await dispatch(createFollowedShow(days, id));
    this.clearSnackbar();
  };

  clearSnackbar = () => {
    this.setState({ isOpen: false });
  };

  handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }
    this.clearSnackbar();
  };

  setUndoData = (show) => {
    this.queue.push({
      show,
      key: moment().valueOf()
    });
  };

  processQueue = () => {
    if (this.queue.length) {
      this.setState({
        isOpen: true,
        undoData: this.queue.shift()
      });
    }
  };

  render() {
    return (
      <View
        {...this.props}
        {...this.state}
        days={Object.values(routes)}
        handleChange={this.handleChange}
        handleSwipeChange={this.handleSwipeChange}
        unFollow={this.unFollow}
        undoAction={this.undoAction}
        handleClose={this.handleClose}
        handleSnackbarExit={this.processQueue}
      />
    );
  }
}

const mapStateToProps = ({ scroll: { directionDown } }) => ({
  directionDown
});

export default connect(mapStateToProps)(MyShows);
