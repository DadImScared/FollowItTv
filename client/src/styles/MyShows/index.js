
import View from './View.css';

export {
  View
};
