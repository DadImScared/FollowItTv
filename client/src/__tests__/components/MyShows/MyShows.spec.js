
import React from 'react';

import axios from 'axios';
import { shallow } from 'enzyme';
import moment from 'moment';
import sinon from 'sinon';

import { MyShows } from '../../../components/MyShows/index';

jest.mock('axios');

describe('MyShows', () => {
  const _date = Date.now;
  let props;

  beforeEach(() => {
    props = {
      followedShows: {},
      followedShowsById: [],
      shows: {},
      location: {
        pathname: '/my_shows/All'
      },
      dispatch: sinon.spy()
    };
    moment.now = () => 1531716593863;
    axios.get.mockImplementation(() => ({ data: [] }));
    axios.post.mockImplementation(() => ({ data: [] }));
  });

  afterEach(() => {
    moment.now = _date;
  });

  test('unFollow', async () => {
    const wrapper = shallow(<MyShows {...props} />);
    const instance = wrapper.instance();
    await instance.unFollow({ schedule: { days: ['Monday'] }, id: 1 });
    expect(props.dispatch.calledOnce).toEqual(true);
    expect(instance.state).toMatchSnapshot('un follow while snackbar is closed');
    await instance.unFollow({ schedule: { days: ['Monday'] }, id: 1 });
    expect(instance.state).toMatchSnapshot('un follow while snackbar is open');
  });

  test('undoAction', async () => {
    const wrapper = shallow(<MyShows {...props} />);
    wrapper.setState({
      isOpen: true,
      undoData: {
        key: 0,
        show: { schedule: { days: ['Monday'] }, id: 1 }
      }
    });
    const instance = wrapper.instance();
    await instance.undoAction();
    expect(props.dispatch.calledOnce).toEqual(true);
    expect(instance.state).toMatchSnapshot();
  });

  test('setInitialTab', () => {
    const newProps = {
      ...props,
      location: {
        pathname: '/my_shows'
      }
    };
    const wrapper = shallow(<MyShows {...newProps} />);
    const instance = wrapper.instance();
    instance.setInitialTab(newProps);
    expect(instance.state.day).toEqual(0);
  });

  test('setInitialTab with day', () => {
    const newProps = {
      ...props,
      location: {
        pathname: '/my_shows/Monday'
      }
    };
    const wrapper = shallow(<MyShows {...newProps} />);
    const instance = wrapper.instance();
    instance.setInitialTab(newProps);
    expect(instance.state.day).toEqual(1);
  });
});
