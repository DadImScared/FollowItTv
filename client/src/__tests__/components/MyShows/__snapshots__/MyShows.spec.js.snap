// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyShows undoAction 1`] = `
Object {
  "day": 0,
  "isOpen": false,
  "undoData": Object {
    "key": 0,
    "show": Object {
      "id": 1,
      "schedule": Object {
        "days": Array [
          "Monday",
        ],
      },
    },
  },
}
`;

exports[`un follow while snackbar is closed 1`] = `
Object {
  "day": 0,
  "isOpen": true,
  "undoData": Object {
    "key": 1531716593863,
    "show": Object {
      "id": 1,
      "schedule": Object {
        "days": Array [
          "Monday",
        ],
      },
    },
  },
}
`;

exports[`un follow while snackbar is open 1`] = `
Object {
  "day": 0,
  "isOpen": false,
  "undoData": Object {
    "key": 1531716593863,
    "show": Object {
      "id": 1,
      "schedule": Object {
        "days": Array [
          "Monday",
        ],
      },
    },
  },
}
`;
